### PR TITLE
Update token pool spec

### LINKED ISSUE #206: Begin K spec for `update_token_pool`
We'll start with a version that does not use #207 to make progress, then we'll migrate to using the `<operations>` cell as well.

### PR DESCRIPTION
**Pull request by hjorthjort**
Closes #206 

I have a working positive case for the `update_token_pool` function but I have some question marks. I'm not sure exactly why some of these initial values need to look the way they look, or if the way I've written the spec is sound. Help appreciated.

### CODE REVIEW

File: tests/proofs/dexter/dexter-spec.md [hunk 1]
```diff
@@ -122,6 +123,39 @@ If any of the conditions are not satisfied, the call fails.
...
+  claim <k> #runProof(IsFA2, UpdateTokenPool) => . </k>
+        <stack> .Stack </stack>
+        <selfIsUpdatingTokenPool> false => true </selfIsUpdatingTokenPool>
+        <tokenAddress> TokenAddress:Address </tokenAddress>
+
+        <myaddr> SelfAddress </myaddr>
+        <myamount> #Mutez(Amount) </myamount>
+        <senderaddr> Sender </senderaddr>
+        <sourceaddr> Sender </sourceaddr>
+        <paramtype> #Type(nat) </paramtype>  // Why does this work????? Why would this be the type???
```
Comment (sskeirik) [inline tests/proofs/dexter/dexter-spec.md:149]: This has to do with the fact that our semantics has never correctly handled a concept called `entrypoints`; unfortunately, this contract uses the entrypoints feature. What happens is: the contract has a certain parameter type.  Typically, when you invoke a contract, you pass in some value of the parameter type.

However, when the parameter type is a nested `or` type, we can label the leaves of the `or` with an entrypoint annotation. When you invoke a contract with an entrypoint, you don't need to mention the nested `or` part ---  you just need the entrypoint name and the leaf type.

In this case, the contract is looking up the type of its `update_token_pool_internal` entrypoint is --- (which is `nat`). However, the contract does not actually check that this entrypoint has any type in particular. It just fowards this information along to the contract that it later inovkes as a callback.

File: tests/proofs/dexter/dexter-spec.md [hunk 2]
```diff
@@ -122,6 +123,39 @@ If any of the conditions are not satisfied, the call fails.
...
+        <knownaddrs> KnownAddresses </knownaddrs>
+        <returncode> 111 => ?_ </returncode> //  This is necessary if we hit an #Assume(false), I think that means we're on an "impossible" branch? So it's correct to include it?
```
Comment (sskeirik) [inline tests/proofs/dexter/dexter-spec.md:151]: This should not be necessary, as far as I know. My understanding is: if we can reach a point where the returncode changed, we can reach a point where the return code has not changed and all other parts of the state are identical.

File: tests/proofs/dexter/dexter-spec.md [hunk 3]
```diff
@@ -122,6 +123,39 @@ If any of the conditions are not satisfied, the call fails.
...
+        <operations> _ => [ Transfer_tokens Pair SelfAddress #Contract ( SelfAddress , nat ) #Mutez (0) TokenAddress ?_ ] ;; .InternalList </operations>
+        <nonce> #Nonce(O) => #Nonce(O +Int 1) </nonce>
+    requires Amount ==Int 0
+     andBool notBool IsFA2
+     andBool TokenAddress in_keys(KnownAddresses)
+     andBool KnownAddresses[TokenAddress] ==K #Contract(TokenAddress, #Type(pair address (contract nat)))
+     andBool KnownAddresses[SelfAddress] ==K #Contract(SelfAddress, SelfType)
```
Comment (sskeirik) [inline tests/proofs/dexter/dexter-spec.md:158]:
```suggestion
        <paramtype> #Type(nat) </paramtype>  // NOTE: return type of `update_token_pool_internal` entrypoint
        <knownaddrs> KnownAddresses </knownaddrs>
        <operations> _ => [ Transfer_tokens Pair SelfAddress #Contract ( SelfAddress , nat ) #Mutez(0) TokenAddress ?_ ] ;; .InternalList </operations>
        <nonce> #Nonce(O) => #Nonce(O +Int 1) </nonce>
    requires Amount ==Int 0
     andBool notBool IsFA2
     andBool TokenAddress in_keys(KnownAddresses)
     andBool KnownAddresses[TokenAddress] ==K #Contract(TokenAddress, #Type(pair address (contract nat)))
```

Comment (sskeirik) [reply tests/proofs/dexter/dexter-spec.md:158]: The `<returncode>` cell modification was not needed as far as I could tell --- at least, I applied the same diff to my local codebase (modulo any typos I made) and it worked.

File: tests/proofs/dexter/dexter-spec.md (hunk 1 above)
Comment (hjorthjort) [reply tests/proofs/dexter/dexter-spec.md:149]: So why would we need to have this type set when we are calling `update_token_pool`? I'd understand if we were calling `update_token_pool_internal`. Does this mean we can't handle more then one callback?

File: tests/proofs/dexter/dexter-spec.md (hunk 2 above)
Comment (hjorthjort) [reply tests/proofs/dexter/dexter-spec.md:151]: well, this is to do with the fact that we hit an `#Assume(false)` it seems, and that only rewrites to `.` by rewriting the return-code, right?

File: tests/proofs/dexter/dexter-spec.md (hunk 3 above)
Comment (hjorthjort) [reply tests/proofs/dexter/dexter-spec.md:158]: Great, thanks!

**Comment by sskeirik** [comment]
@hjorthjort Hey Rikard, I noticed you commented out all of the other proofs?

File: tests/proofs/dexter/dexter-spec.md (hunk 2 above)
Comment (sskeirik) [reply tests/proofs/dexter/dexter-spec.md]: Oh, you're right. This rule:

```
rule <k> #Assume(false) ... </k>
     <returncode> 111 => 0 </returncode>
```

was added recently. This rule is no longer necessary; it should be deleted.

**Comment by hjorthjort** [comment]
> @hjorthjort Hey Rikard, I noticed you commented out all of the other proofs?

Yeah, I tend to do that while figuring out a specific one, to make it faster to run.

File: tests/proofs/dexter/dexter-spec.md [hunk 4]
```diff
@@ -124,6 +124,75 @@ If any of the conditions are not satisfied, the call fails.
...
+The contract queries its underlying token contract for its own token balance if the following conditions are satisfied:
+
+1.  the token pool is _not_ currently updating (i.e. `storage.selfIsUpdatingTokenPool = false`)
+2.  exactly 0 tez was transferred to this contract when it was invoked
+3.  the txn sender must be equal to the txn source
+4.  if we are running the FA2 version of Dexter, then check that the contract at address `storage.tokenAddress` has a well-typed FA2 `balance_of` entrypoint;
+    otherwise, check that the contract at address `storage.tokenAddress` has a well-typed FA12 `get_balance` entrypoint.
+
+- TODO: Combine these positive cases into one spec.
+- TODO: Negative case
```
Comment (sskeirik) [inline tests/proofs/dexter/dexter-spec.md:138]:
```suggestion
```

It seems these comments were already taken care of. Is that correct?

**Review by sskeirik — APPROVED** [review]
Overall, looks good to me. The only thing that surprises me is that you needed to split the negative case into two subproofs. Was that for correctness or performance?

**Comment by hjorthjort** [comment]
> Overall, looks good to me. The only thing that surprises me is that you needed to split the negative case into two subproofs. Was that for correctness or performance?

Performance. I got "broken pipe" issues without it
I'll try pushing the combined spec and see if CI handles it better

**Comment by hjorthjort** [comment]
Okay, latest update: it's not just performance. Seems the prover can't resolve some of the `orBool` conditions into "either A or B or both", but rather gets stuck with something like this on the path conditions:
```
   true          
  #Equals                                                                                                                                                                     notBool ( Sender:Address ~> . ) ==K ( Source:Address ~> . ) orBool ( ( KnownAddresses [ TokenAddress:Address ] ~> . ) ==K ( #Contract ( _A , T ) ~> . ) andBool notBoo
l ( T:Type ~> . ) ==K ( pair .AnnotationList address .AnnotationList contract .AnnotationList nat .AnnotationList ~> . ) )
```

Let's use the double claims for now. I'll spend some time trying to resolve this in order to improve on #234, but it's already turning into a time sink to merge the two failure claims and I'd rather move on.

**Pull request merged**